File: RemixedDungeon/src/main/java/com/watabou/pixeldungeon/actors/buffs/ManaRegeneration.java
package com.watabou.pixeldungeon.actors.buffs;

import com.watabou.pixeldungeon.Dungeon;
import com.watabou.pixeldungeon.Facilitations;
import com.watabou.pixeldungeon.actors.Char;

import org.jetbrains.annotations.NotNull;

public class ManaRegeneration extends Buff {

    private static final float REGENERATION_DELAY = 20;

    @Override
    public boolean act() {
        if (target.isAlive()) {
            if (!target.level().isSafe()) {
                target.accumulateSkillPoints(1);
            }

            final int[] bonus = {0};

            if(Dungeon.isFacilitated(Facilitations.FAST_MANA_REGENERATION)) {
                bonus[0] += 10;
            }

            target.forEachBuff(b-> bonus[0] +=b.manaRegenerationBonus());

            spend((float) (REGENERATION_DELAY / Math.pow(1.2, bonus[0])));
        } else {
            deactivateActor();
        }
        return true;
    }

	@Override
	public boolean attachTo(@NotNull Char target ) {
        return target.buffLevel(getEntityKind()) > 0 || super.attachTo(target);
    }
}

File: scripts/buffs/ManaRegeneration.lua
---
--- ManaRegeneration: skill point tick, rate scaled by manaRegenerationBonus
--- buffs (was actors/buffs/ManaRegeneration.java)
---
local RPD  = require "scripts/lib/commonClasses"
local buff = require "scripts/lib/buff"

local Facilitations = luajava.bindClass("com.watabou.pixeldungeon.Facilitations")

local REGENERATION_DELAY = 20

return buff.init{
    desc  = function ()
        return {
            icon          = -1, -- no indicator icon
            name          = "ManaRegenerationBuff_Name",
            info          = "ManaRegenerationBuff_Info",
        }
    end,

    act = function(self, buff)
        local target = buff.target
        if not target:isAlive() then
            buff:deactivateActor()
            return
        end

        if not target:level():isSafe() then
            target:accumulateSkillPoints(1)
        end

        local bonus = target:manaRegenerationBonusSum()
        if RPD.Dungeon:isFacilitated(Facilitations.FAST_MANA_REGENERATION) then
            bonus = bonus + 10
        end

        buff:spend(REGENERATION_DELAY / math.pow(1.2, bonus))
    end
}
